mpqp/tools/theoretical_simulation.py

```diff
@@ -0,0 +1,210 @@
+from __future__ import annotations
+
+import numpy as np
+import numpy.typing as npt
+from matplotlib import pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
+from scipy.spatial.distance import jensenshannon
+
+from mpqp import QCircuit
+from mpqp.core.instruction.measurement.basis_measure import BasisMeasure
+from mpqp.execution import AWSDevice
+from mpqp.execution.devices import AvailableDevice
+from mpqp.execution.runner import _run_single  # pyright: ignore[reportPrivateUsage]
+from mpqp.noise import Depolarizing
+from mpqp.noise.noise_model import NoiseModel
+
+
+def theoretical_probs(
+    circ: QCircuit,
+) -> npt.NDArray[np.float32]:
+    """Computes the theoretical probabilities of a (potentially noisy circuit
+    execution.
+
+    Args:
+        circ: The circuit to run.
+
+    Returns:
+        The probabilities corresponding to each basis state.
+    """
+    d: int = 2 ** (circ.nb_qubits)
+
+    state = np.zeros((d, d), dtype=np.complex64)
+    state[0, 0] = 1
+
+    for gate in circ.get_gates():
+        g = gate.to_matrix(circ.nb_qubits).astype(np.complex64)
+        state = g @ state @ g.T.conj()
+        for noise in circ.noises:
+            if (
+                len(noise.gates) == 0
+                or type(gate) in noise.gates
+                and gate.connections().issubset(noise.targets)
+            ):
+                state = sum(
+                    (
+                        k @ state @ k.T.conj()
+                        for k in noise.to_adjusted_kraus_operators(
+                            gate.connections(), circ.nb_qubits
+                        )
+                    ),
+                    start=np.zeros((d, d), dtype=np.complex64),
+                )
+
+    connected_qubits = set().union(*[gate.connections() for gate in circ.get_gates()])
+    unconnected_qubits = set(range(circ.nb_qubits)).difference(connected_qubits)
+    for noise in circ.noises:
+        if len(noise.gates) == 0:
+            sum(
+                (
+                    k @ state @ k.T.conj()
+                    for k in noise.to_adjusted_kraus_operators(
+                        unconnected_qubits, circ.nb_qubits
+                    )
+                ),
+                start=np.zeros((d, d), dtype=np.complex64),
+            )
+
+    return state.diagonal().real
+
+
+def dist_alpha_matching(alpha: float):
+    """The trust interval is computed from the distance between the circuit
+    without noise and the noisy circuits probability distributions. This
+    interval depends on this distance in a non linear manner. This function
+    gives the relation between the two.
+
+    Args:
+        alpha: The distance Jensen-Shannon between the non noisy distribution
+            and the noisy distribution.
+
+    Returns:
+        The trust interval for the distance.
```

Comment:
We don't really return here an interval, but rather the size of the trust interval, right?

Reply:
nitpicking much ? :p

Reply:
Not really ^^, I struggled a bit to catch the differences between the intervals, the several distances, the interval size etc. Maybe we can detail that in the onlline documentation, to put some theoretical context?

Reply:
good point, this should go in the header of this file I guess

Reply:
I added a big ol piece of documentation, I hope it will please you (for my standards, I was exceptionally verbose! You were right, it was needed)